src/subject.rs

```diff
@@ -0,0 +1,153 @@
+    /// Unique identifier for user
+    #[serde(rename(serialize = "uid"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub user_id: Option<String>,
+
+    /// Time zone of client device's OS
+    ///
+    /// This field is set automatically in enrichment if not present
+    #[serde(rename(serialize = "tz"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub timezone: Option<String>,
+
+    /// User language
+    #[serde(rename(serialize = "lang"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub language: Option<String>,
+
+    /// User IP
+    ///
+    /// This field is set automatically in enrichment if not present
+    #[serde(rename(serialize = "ip"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub ip_address: Option<String>,
+
+    /// Useragent
+    ///
+    /// This field is set automatically in enrichment if not present
+    #[serde(rename(serialize = "ua"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub useragent: Option<String>,
+
+    /// Domain-specific unique user identifier
+    #[serde(rename(serialize = "duid"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub domain_uid: Option<String>,
+
+    /// Used to overwrite `nuid`
+    /// https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/snowplow-tracker-protocol/#user-related-parameters
+    ///
+    /// This field is set automatically in enrichment if not present
+    #[serde(rename(serialize = "tnuid"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub network_uid: Option<String>,
+
+    /// Unique identifier (UUID) for this visit of this user_id to this domain
+    #[serde(rename(serialize = "sid"))]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub session_uid: Option<Uuid>,
+}
+
+impl Subject {
+    pub fn builder() -> SubjectBuilder {
+        SubjectBuilder::default()
+    }
+
+    /// Merges another instance of [Subject], with self taking priority
+    pub(crate) fn merge(self, other: Subject) -> Self {
+        Self {
+            user_id: self.user_id.or(other.user_id),
+            timezone: self.timezone.or(other.timezone),
+            language: self.language.or(other.language),
+            ip_address: self.ip_address.or(other.ip_address),
+            useragent: self.useragent.or(other.useragent),
+            domain_uid: self.domain_uid.or(other.domain_uid),
+            network_uid: self.network_uid.or(other.network_uid),
+            session_uid: self.session_uid.or(other.session_uid),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_build_subject_all_fields() {
+        let session_uid = Uuid::new_v4();
+        let subject = Subject::builder()
+            .user_id("user_1")
+            .timezone("Europe/London")
+            .language("en")
+            .ip_address("0.0.0.0")
+            .useragent("Mozilla/Firefox")
+            .domain_uid("1")
```

Comment:
Nit: I think it's not enforced in the validation, but usually the domain and network userid are UUIDs. It doesn't matter here, just in case someone was taking inspiration from the tests, might be better to use the UUID format.